tensorboard/data/server/run.rs

```diff
@@ -99,15 +102,30 @@ struct StageValue {
 }
 
 impl StageTimeSeries {
-    fn new(metadata: Box<pb::SummaryMetadata>) -> Self {
+    fn new(
+        metadata: Box<pb::SummaryMetadata>,
+        plugin_sampling_hint: Arc<Option<PluginSamplingHint>>,
+    ) -> Self {
         let data_class =
             pb::DataClass::from_i32(metadata.data_class).unwrap_or(pb::DataClass::Unknown);
-        let capacity = match data_class {
+        let mut capacity = match data_class {
             pb::DataClass::Scalar => 1000,
             pb::DataClass::Tensor => 100,
             pb::DataClass::BlobSequence => 10,
             _ => 0,
         };
+        // Override the capacity using the plugin-specific hint.
+        if capacity > 0 {
+            if let Some(ref pd) = metadata.plugin_data {
+                if let Some(hint) = &*plugin_sampling_hint {
+                    if let Some(&num_samples) = hint.0.get(&pd.plugin_name) {
+                        // TODO(psybuzz): if the hint prescribes 0 samples, the reservoir should ideally
+                        // be unbounded. For now, it simply creates a reservoir with capacity 0.
```

Comment:
Two questions to simplify this call site:

 1. From this logic, it looks like a `PluginSamplingHint` with an empty
    hash map serves as a nice “zero value”. What do you think about
    making the plugin sampling hint required to remove one layer of
    conditionals?

 2. What do you think about encapsulating this logic into a method on
    `PluginSamplingHint`? Maybe:
    
    ```rust
    /// Gets the capacity for a given time series.
    pub fn capacity(&self, metadata: &pb::SummaryMetadata) -> usize;
    // ^ would include the default-values logic
    ```

    In some ways, this just moves the code around, but I think that you
    might find that the implementation simplifies, too. For instance, it
    enables you to just `return 0` from the `match` clause for `_`.

Reply:
> What do you think about making the plugin sampling hint required to remove one layer of conditionals?

Sounds good!

> What do you think about encapsulating this logic into a method on `PluginSamplingHint`?

Hm, if we extracted the logic into a method that accepted `metadata`, it would be able to compute the capacity entirely, so there would be less reason to keep the `DataClass` `match` around in `StageTimeSeries::new`.  In that case, `PluginSamplingHint::capacity(metadata)` should contain all the capacity logic.

If the concern is that the capacity logic is a bit much to have in the constructor, would it suffice to just move the code into a method `StageTimeSeries::capacity`?

aside: also switched the `if capacity > 0` to `if data_class != Unknown`, since even if a known data class had 0 default capacity, the user hint should still override.

Reply:
> Hm, if we extracted the logic into a method that accepted `metadata`, it would be able to compute the capacity entirely, so there would be less reason to keep the `DataClass` `match` around in `StageTimeSeries::new`. In that case, `PluginSamplingHint::capacity(metadata)` should contain all the capacity logic.

Agreed.

> If the concern is that the capacity logic is a bit much to have in the constructor, would it suffice to just move the code into a method `StageTimeSeries::capacity`?

I’m not too worried by the number of lines of code in the constructor.
I was wondering whether it makes sense from a conceptual standpoint that
the plugin sampling hint should know how to determine the capacity. For
instance, that would provide a natural place to unit-test that logic if
we so desired.

That said, I don’t feel strongly about it; the question was genuine.
This is fine with me, and I do think that having it in a separate
function (as you’ve done) is nice, because you can return early.

Reply:
Gotcha.  In that case, keeping the logic here in `StageTimeSeries` still seems preferable than moving to `PluginSamplingHint::capacity` imo, so I'll keep as-is for now.